libmacaroons.go

```diff
@@ -4,85 +4,169 @@
+		return m, err
 	}
 	return m, nil
 }
 
-func (m libMacaroon) Bind(primary Macaroon) (Macaroon, error) {
-	m1, err := primary.(libMacaroon).PrepareForRequest(m.Macaroon)
-	if err != nil {
+func (p libMacaroonsPkg) UnmarshalBinary(data []byte) (Macaroon, error) {
+	return nil, fmt.Errorf("unimplemented")
+}
+
+type libMacaroon struct {
+	p    libMacaroonsPkg
+	name string
+}
+
+func (m *libMacaroon) MarshalJSON() ([]byte, error) {
+	expr := fmt.Sprintf(`result = %s.serialize(format='2j')`, m.name)
+	var r string
+	if err := m.p.eval(expr, &r); err != nil {
 		return nil, err
 	}
-	return libMacaroon{m1}, nil
+	return []byte(r), nil
 }
 
-func (m libMacaroon) Verify(rootKey []byte, check Checker, discharges []Macaroon) error {
-	discharges1 := make([]*macaroons.Macaroon, len(discharges))
-	for i, m := range discharges {
-		discharges1[i] = m.(libMacaroon).Macaroon
-	}
-	v := macaroons.NewVerifier()
-	if err := v.SatisfyGeneral(func(caveat string) bool {
-		return check[caveat]
-	}); err != nil {
-		return err
-	}
-	return v.Verify(m.Macaroon, string(rootKey), discharges1...)
+func (m *libMacaroon) MarshalBinary() ([]byte, error) {
+	return nil, fmt.Errorf("unimplemented")
 }
 
-type libMacaroonPkg struct{}
+func (m *libMacaroon) WithFirstPartyCaveat(caveatId string) (Macaroon, error) {
+	m1 := m.p.newMacaroon()
+	expr := fmt.Sprintf(`%s = %s.add_first_party_caveat(%s)`,
+		m1.name, m.name, pyVal(caveatId))
+	if err := m.p.eval(expr, nil); err != nil {
+		return nil, err
+	}
+	return m1, nil
+}
 
-func (libMacaroonPkg) UnmarshalJSON(data []byte) (Macaroon, error) {
-	var m macaroons.Macaroon
-	if err := m.UnmarshalJSON(data); err != nil {
+func (m *libMacaroon) WithThirdPartyCaveat(rootKey []byte, caveatId string, loc string) (Macaroon, error) {
+	m1 := m.p.newMacaroon()
+	// TODO specify nonce explicitly.
+	expr := fmt.Sprintf(`%s = %s.add_third_party_caveat(%s, %s, %s)`,
+		m1.name, m.name, pyVal(loc), pyVal(rootKey), pyVal(caveatId))
+	if err := m.p.eval(expr, nil); err != nil {
 		return nil, err
 	}
-	return libMacaroon{&m}, nil
+	return m1, nil
 }
 
-func (libMacaroonPkg) UnmarshalBinary(data []byte) (Macaroon, error) {
-	var m macaroons.Macaroon
-	if err := m.UnmarshalBinary(data); err != nil {
+func (m *libMacaroon) Bind(primary Macaroon) (Macaroon, error) {
+	pm := primary.(*libMacaroon)
+	boundM := m.p.newMacaroon()
+	expr := fmt.Sprintf(`%s = %s.prepare_for_request(%s)`, boundM.name, pm.name, m.name)
+	if err := m.p.eval(expr, nil); err != nil {
 		return nil, err
 	}
-	return libMacaroon{&m}, nil
+	return boundM, nil
 }
 
-func (libMacaroonPkg) New(rootKey []byte, id, loc string) (Macaroon, error) {
-	m, err := macaroons.NewMacaroon(loc, string(rootKey), id)
+func (m *libMacaroon) Verify(rootKey []byte, check Checker, discharges []Macaroon) error {
+	dischargeNames := make([]string, len(discharges))
+	for i, m := range discharges {
+		dischargeNames[i] = m.(*libMacaroon).name
+	}
+	checkData, _ := json.Marshal(check)
```

Comment:
panic if err ?

Reply:
it's a map[string]bool - it can't fail. (and if it does, the tests'll fail too)